File: iOS/Tool/Thead/AliveThread.m
//
//  AliveThread.m
//  OCDemo
//
//  Created by Felix on 2023/1/21.
//

#import "AliveThread.h"

@interface LiveThread : NSThread
@end
@implementation LiveThread

- (void)dealloc {
    NSLog(@"%s", __func__);
}

@end

@interface AliveThread ()

@property (nonatomic,strong) LiveThread *liveThread;
@property (nonatomic, assign) BOOL stopped;

@end

@implementation AliveThread

- (instancetype)init {
    if (self = [super init]) {
        self.stopped = NO;
        
        __weak typeof(self) weakSelf = self;
        self.liveThread = [[LiveThread alloc] initWithBlock:^{
            NSLog(@"------常驻线程开启-------");
            [[NSRunLoop currentRunLoop] addPort:[[NSPort alloc] init] forMode:NSDefaultRunLoopMode];
            
            while (weakSelf && !weakSelf.stopped) {
                [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate distantFuture]];
            }
            NSLog(@"------常驻线程结束------");
        }];
        [self.liveThread start];
    }
    return self;
}


- (void)executeTask:(AliveThreadTask)task {
    if (!self.liveThread || !task) return;
    
    [self performSelector:@selector(realizeTask:) onThread:self.liveThread withObject:task waitUntilDone:NO];
}

- (void)realizeTask:(AliveThreadTask)task {
    task();
}

- (void)stop {
    if (!self.liveThread) return;
    [self performSelector:@selector(liveThreadStop) onThread:self.liveThread withObject:nil waitUntilDone:YES];
}


- (void)liveThreadStop{
    self.stopped = YES;
    CFRunLoopStop(CFRunLoopGetCurrent());
    self.liveThread = nil;
}

- (void)dealloc {
    NSLog(@"%s", __func__);
    
    [self stop];
}

@end

File: XFlash_iOS/XFlash/Classes/Public/AliveThread/AliveThread.swift
//
//  AliveThread.swift
//  Pods-XFlash_Example
//
//  Created by Felix on 2023/7/31.
//

import Foundation

public typealias VoidCallback = () -> Void
public class LiveThread: Thread {
    
    deinit {
        print("LiveThread-deinit")
    }
}

@objcMembers
public class AliveThread: NSObject {
    private(set)
    public var liveThread: LiveThread?
    
    var stopped: Bool = false
    var task: VoidCallback?
    
    // MARK: - LifeCycle

    public override init() {
        super.init()
        
        liveThread = LiveThread(block: { [weak self] in
            guard let self = self else { return }
            print("------常驻线程开启-------")
            
            RunLoop.current.add(Port(), forMode: .default)
            while !self.stopped {
                RunLoop.current.run(mode: .default, before: Date.distantFuture)
            }
            
            print("------常驻线程结束-------")
        })
        liveThread?.start()
    }
    
    deinit {
        print("AliveThread-deinit")
    }
}

// MARK: - Public
public extension AliveThread {
    
    func execute(with task: @escaping VoidCallback) {
        guard let liveThread = liveThread else { return }
        
        self.task = task
        perform(#selector(realizeTask),
                on: liveThread,
                with: nil,
                waitUntilDone: false)
    }
    
    func stop() {
        guard let liveThread = liveThread else { return }
        
        perform(#selector(stopLiveThread),
                on: liveThread,
                with: nil,
                waitUntilDone: true)
    }
}


private extension AliveThread {
    
    @objc func realizeTask() {
        guard let task = task else { return }
        task()
    }
    
    @objc func stopLiveThread() {
        stopped = true
        CFRunLoopStop(CFRunLoopGetCurrent())
        liveThread = nil
    }
}
